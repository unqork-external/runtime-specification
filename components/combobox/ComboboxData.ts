import { Description, Optional, Required } from '@tsed/schema'

import { ComboboxOption } from './ComboboxOption'
import { ComboboxRefreshTarget } from './ComboboxRefreshTarget'

export class ComboboxData {
  @Required()
  @Description('Property ID of the referenced component.')
  options: ComboboxOption[]

  @Optional()
  @Description(
    'Property map indicating which properties in the data reference component value array map to in a Combobox option.',
  )
  refreshTarget?: ComboboxRefreshTarget
}
