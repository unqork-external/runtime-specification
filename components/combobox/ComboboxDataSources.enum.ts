export enum ComboboxDataSources {
  'local' = 'local',
  'reference' = 'reference',
}
