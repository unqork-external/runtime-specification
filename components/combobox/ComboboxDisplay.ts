import { Default, Description, Optional } from '@tsed/schema'

import { Display } from '../../component-composition'

export class ComboboxDisplay extends Display {
  @Optional()
  @Description('Whether to show a loading state instead of "No Options". Useful when loading data externally.')
  @Default(false)
  loading? = false
}
