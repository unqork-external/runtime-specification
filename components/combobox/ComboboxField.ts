import { Default, Description, Optional } from '@tsed/schema'

import { Field, InputHelperText, InputLabel } from '../../component-composition'

export class ComboboxField extends Field {
  @Optional()
  @Description('Displays a clear button at the leading of the component.')
  @Default(true)
  clearable? = true

  @Optional()
  @Description('Label for the clear button used by screen readers.')
  clearButtonLabel?: string

  @Optional()
  @Description('Label for the close popup indicator button used by screen readers.')
  closeButtonLabel?: string

  @Optional()
  @Description('Allows the user to store a custom value not present in the options.')
  @Default(false)
  creatable? = false

  @Optional()
  @Description('Label for the creatable option to be displayed. Replaces `{VALUE}` with the text entered by the user.')
  createOptionLabel?: string

  @Optional()
  @Description("Custom CSS class applied to the component's wrapper HTML element.")
  customClass?: string

  @Optional()
  @Description('Whether filtering is handled by external logic, disabling internal filtering.')
  @Default(false)
  externalFilter? = false

  @Optional()
  @Description('Whether options are grouped by their Group Label.')
  @Default(false)
  groupOptions? = false

  @Optional()
  @Description('Helper text displayed below the input control.')
  helperText: InputHelperText = new InputHelperText()

  @Optional()
  @Description('Label for the input control.')
  label: InputLabel = new InputLabel()

  @Optional()
  @Description('Label displayed in the options list when `display.loading` is `true` and there are no options.')
  loadingLabel?: string

  @Optional()
  @Description('Allows multiple options selection.')
  @Default(false)
  multiple? = false

  @Optional()
  @Description('Label displayed when there are no options. Replaces `{VALUE}` with the text entered by the user.')
  noOptionsLabel?: string

  @Optional()
  @Description('Label for the open popup indicator button used by screen readers.')
  openButtonLabel?: string

  @Optional()
  @Description('Text displayed as a placeholder label when there are no options selected.')
  placeholder?: string

  @Optional()
  @Description(
    'Whether the options popup should remain open after a selection is made. Useful when multiple selections are allowed.',
  )
  @Default(false)
  remainOpenAfterSelection? = false

  @Optional()
  @Description('Whether the selected options should be removed from the options list.')
  @Default(false)
  removeSelectedOptions? = false

  @Optional()
  @Description('Whether the leading icon on the component should be replaced by the icon of the selected option.')
  @Default(false)
  replaceIconWithSelectedOptionIcon? = false

  @Optional()
  @Description('Whether the selected option icon should appear outside of the options list.')
  @Default(false)
  showSelectedOptionIcon? = false

  @Optional()
  @Description("Text for the component's tooltip.")
  tooltipDescription?: string

  @Optional()
  @Description(
    'Whether to enable the typeahead behavior on the search text input as the user selects options with the keyboard.',
  )
  @Default(false)
  typeahead? = false
}
