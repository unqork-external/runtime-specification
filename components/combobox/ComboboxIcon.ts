import { Description, Enum, Required } from '@tsed/schema'

import { TrimmedDescription } from '../../../decorators'
import { IconSource } from '../icon'

export class ComboboxIcon {
  @Enum(IconSource)
  @Required()
  @Description('sourceType determines how the icon is sourced. If url, a valid url must also be defined in source.')
  sourceType: IconSource

  @Required()
  @TrimmedDescription(`
    String representing the source of the icon.
    For url, this should be a fully qualified url to the hosted icon image.
    For fontAwesome, this should be a string describing the fontAwesome classes.
    For cssIcon, this string should be the css class definitions to apply.
  `)
  source: string
}
