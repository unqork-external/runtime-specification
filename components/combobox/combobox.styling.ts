import { Description, Property } from '@tsed/schema'

import { ComboboxTargets } from './comboboxTargets.enum'
import { Css } from '../../../decorators/schema/css.decorator'
import { type CssClassOrProps } from '../../styling/cssClassOrProps.type'
import { StylingModel } from '../../styling/styling.model'
import { TargetsModel } from '../../viewTargets/targets.model'

export class ComboboxStyleTargets implements TargetsModel<ComboboxTargets> {
  @Css()
  @Description('Clear button styles.')
  [ComboboxTargets.clearButton]: CssClassOrProps;

  @Css()
  @Description('Control styles.')
  [ComboboxTargets.control]: CssClassOrProps;

  @Css()
  @Description('Error wrapper styles.')
  [ComboboxTargets.errorsFieldError]: CssClassOrProps;

  @Css()
  @Description('Error icon wrapper styles when in simple view.')
  [ComboboxTargets.errorsIcon]: CssClassOrProps;

  @Css()
  @Description('Error tooltip styles when in simple view.')
  [ComboboxTargets.errorsTooltip]: CssClassOrProps;

  @Css()
  @Description('Helper text styles.')
  [ComboboxTargets.helperText]: CssClassOrProps;

  @Css()
  @Description('Icon container styles.')
  [ComboboxTargets.icon]: CssClassOrProps;

  @Css()
  @Description('Label styles.')
  [ComboboxTargets.label]: CssClassOrProps;

  @Css()
  @Description('Loading label styles.')
  [ComboboxTargets.loadingLabel]: CssClassOrProps;

  @Css()
  @Description('No options label styles.')
  [ComboboxTargets.noOptionsLabel]: CssClassOrProps;

  @Css()
  @Description('Option description styles.')
  [ComboboxTargets.optionDescription]: CssClassOrProps;

  @Css()
  @Description('Option group list item styles.')
  [ComboboxTargets.optionGroupItem]: CssClassOrProps;

  @Css()
  @Description('Option icon container styles.')
  [ComboboxTargets.optionIcon]: CssClassOrProps;

  @Css()
  @Description('Option item styles.')
  [ComboboxTargets.optionItem]: CssClassOrProps;

  @Css()
  @Description('Option label styles.')
  [ComboboxTargets.optionLabel]: CssClassOrProps;

  @Css()
  @Description('Option text styles.')
  [ComboboxTargets.optionText]: CssClassOrProps;

  @Css()
  @Description('Popover wrapper styles.')
  [ComboboxTargets.popover]: CssClassOrProps;

  @Css()
  @Description('Popup indicator button styles.')
  [ComboboxTargets.popupIndicatorButton]: CssClassOrProps;

  @Css()
  @Description('Read only styles.')
  [ComboboxTargets.readOnly]: CssClassOrProps;

  @Css()
  @Description('Wrapper styles.')
  [ComboboxTargets.root]: CssClassOrProps;

  @Css()
  @Description('Search input styles.')
  [ComboboxTargets.searchInput]: CssClassOrProps;

  @Css()
  @Description('Selected option tag styles (when multiple is enabled).')
  [ComboboxTargets.selectedOptionTag]: CssClassOrProps;

  @Css()
  @Description('Selected option tag label delete button styles (when multiple is enabled).')
  [ComboboxTargets.selectedOptionTagDeleteButton]: CssClassOrProps;

  @Css()
  @Description('Selected option tag label text styles (when multiple is enabled).')
  [ComboboxTargets.selectedOptionTagLabel]: CssClassOrProps;

  @Css()
  @Description('Styles for the root component (when in simple view).')
  [ComboboxTargets.simpleView]: CssClassOrProps;

  @Css()
  @Description('Styles for the simple view text mode.')
  [ComboboxTargets.simpleViewText]: CssClassOrProps;

  @Css()
  @Description('Selected option tag styles (when in a simple view).')
  [ComboboxTargets.simpleViewSelectedOption]: CssClassOrProps;

  @Css()
  @Description('Selected option tag label text styles (when in a simple view).')
  [ComboboxTargets.simpleViewSelectedOptionLabel]: CssClassOrProps;

  @Css()
  @Description('Tooltip button styles.')
  [ComboboxTargets.tooltipButton]: CssClassOrProps;

  @Css()
  @Description('Tooltip popover styles.')
  [ComboboxTargets.tooltipPopover]: CssClassOrProps
}

export class ComboboxStyling extends StylingModel<ComboboxStyleTargets> {
  @Property()
  declare targets: ComboboxStyleTargets
}
