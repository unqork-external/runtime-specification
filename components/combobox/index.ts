export * from './ComboboxData'
export * from './ComboboxDataSources.enum'
export * from './ComboboxDisplay'
export * from './ComboboxField'
export * from './ComboboxNestable'
export * from './ComboboxOption'
export * from './ComboboxRefreshTarget'
export * from './combobox.definition'
export * from './combobox.styling'
export * from './comboboxTargets.enum'
