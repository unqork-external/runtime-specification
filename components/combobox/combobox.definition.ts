import { Const, Default, Description, DiscriminatorValue, Optional, Property, Required } from '@tsed/schema'

import { ComboboxStyling } from './combobox.styling'
import { ComboboxData } from './ComboboxData'
import { ComboboxDisplay } from './ComboboxDisplay'
import { ComboboxField } from './ComboboxField'
import { ComboboxIcon } from './ComboboxIcon'
import { ComboboxTargets } from './comboboxTargets.enum'
import { ViewTargets } from '../../../decorators'
import { OneOrMany } from '../../../decorators/schema/oneOrMany.decorator'
import {
  BaseComponentDefinition,
  CommonSimpleViewModel,
  type ComponentWithSimpleView,
} from '../../base-component-interface'
import type { SignalTargets } from '../../signals'

@DiscriminatorValue('combobox')
@ViewTargets(ComboboxTargets)
export class ComboboxComponentDefinition extends BaseComponentDefinition implements ComponentWithSimpleView {
  @Const('combobox')
  type: 'combobox' = 'combobox' as const

  @Required()
  @Description('Data for the Combobox options.')
  data: ComboboxData

  @Required()
  @Description('Display configuration for the Combobox.')
  display: ComboboxDisplay

  @Required()
  @Description('Field configuration for the Combobox.')
  field: ComboboxField

  @Optional()
  @Property(ComboboxIcon)
  @Description('Icon to display on the leading side of the Combobox.')
  icon?: ComboboxIcon

  @Optional()
  @Description('Read-only value of the search input as the user modifies it.')
  searchValue?: string

  @Optional()
  @Description('Simple view configuration.')
  simpleView? = new CommonSimpleViewModel()

  @Optional()
  @Description('Map of customized CSS styling for specific targets.')
  styling?: ComboboxStyling

  @Optional()
  @Description('Map of signals configuration for specific targets.')
  signals?: SignalTargets<ComboboxTargets>

  @Optional()
  @Description('Simple view state.')
  @Default(false)
  useSimpleView? = false

  @Optional()
  @Description('Value for the Combobox. `string` if field.multiple is disabled, `Array<sting>` if it is enabled.')
  @OneOrMany('string')
  value: string | string[]
}
