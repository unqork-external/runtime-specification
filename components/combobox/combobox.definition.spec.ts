import { ComboboxComponentDefinition } from './combobox.definition'
import { generateSchemaAndValidate } from '../../../utilities'

describe('Combobox Contract Tests', function () {
  it('should ensure the contract is clear', function () {
    const { schema } = generateSchemaAndValidate(ComboboxComponentDefinition)

    expect(schema.required).toMatchSnapshot()
    expect(schema.properties).toMatchSnapshot()
  })
})
