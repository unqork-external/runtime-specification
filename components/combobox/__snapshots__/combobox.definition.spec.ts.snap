// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combobox Contract Tests should ensure the contract is clear 1`] = `
[
  "key",
  "type",
  "data",
  "display",
  "field",
]
`;

exports[`Combobox Contract Tests should ensure the contract is clear 2`] = `
{
  "childIds": {
    "description": "Child IDs for all nestable references within the component.",
    "type": "array",
  },
  "data": {
    "$ref": "#/definitions/ComboboxData",
  },
  "display": {
    "$ref": "#/definitions/ComboboxDisplay",
  },
  "executable": {
    "default": true,
    "description": "Defines whether or not the object will respond to events or fire operations. When \`true\`, the object will perform and behave as defined in its spec. When \`false\`, the object will no longer be responsive to the system.",
    "type": "boolean",
  },
  "field": {
    "$ref": "#/definitions/ComboboxField",
  },
  "iconKey": {
    "description": "Key of the icon to display on the leading side of the Combobox, as defined in the \`icons\` nestable.",
    "type": "string",
  },
  "icons": {
    "description": "Children definition for all the icons used in this component.",
    "items": {
      "$ref": "#/definitions/IconComponentDefinition",
    },
    "type": "array",
  },
  "inputs": {
    "items": {
      "$ref": "#/definitions/InputRef",
    },
    "type": "array",
  },
  "key": {
    "description": "'key' is the user-defined, unique identifier for an object within a module. It doesn't have to be unique across different modules. Objects within the runtime use keys to target other objects.",
    "examples": [
      "I am the key",
      "bestKeyEver",
    ],
    "minLength": 1,
    "type": "string",
  },
  "nestables": {
    "description": "Nestable configuration.",
    "type": "object",
  },
  "package": {
    "$ref": "#/definitions/PackageModel",
  },
  "searchValue": {
    "description": "Read-only value of the search input as the user modifies it.",
    "type": "string",
  },
  "signals": {
    "description": "Map of signals configuration for specific targets.",
    "examples": [
      {
        "args": {
          "inputs": {
            "propertyPath": "value",
            "required": true,
            "targetKey": "firstName",
          },
        },
        "id": "firstNameWatcher",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "firstNameWatcher operations fired: value watcher triggered",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onWatch",
      },
      {
        "id": "textfieldOnBlur",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "textfield received blur event from DOM",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onBlur",
      },
      "<Example> Prompt: "I want a textfield that will trigger the root execute signal on myValidationComponent when the textfield input receives a blur event." Response: [{"key":"yourGeneratedTextfield","type":"textfield","signals":{"input":[{"id":"textfieldOnBlur","operations":[{"type":"EXECUTE","options":{"targetKey":"myValidationComponent","signalTarget":"root","shouldExecute":true}}],"type":"onBlur"}]}}] </Example>",
    ],
    "type": "object",
  },
  "simpleView": {
    "description": "Simple view configuration.",
    "type": "object",
  },
  "styling": {
    "$ref": "#/definitions/ComboboxStyling",
  },
  "type": {
    "const": "combobox",
    "description": "Name of the type of component.",
    "examples": [
      "combobox",
    ],
    "minLength": 1,
    "type": "string",
  },
  "useSimpleView": {
    "default": false,
    "description": "Simple view state.",
    "type": "object",
  },
  "validation": {
    "$ref": "#/definitions/Validation",
  },
  "value": {
    "anyOf": [
      {
        "type": "string",
      },
      {
        "items": {
          "type": "string",
        },
        "type": "array",
      },
    ],
    "description": "Value for the Combobox. \`string\` if field.multiple is disabled, \`Array<sting>\` if it is enabled.",
    "type": [
      "null",
      "integer",
      "number",
      "string",
      "boolean",
      "array",
      "object",
    ],
  },
}
`;
